Smarter push/pull fallback from TLS to plaintext

aaronlehmann opened the pull request:
With the --insecure-registry daemon option (or talking to a registry on a local IP), the daemon will first try TLS, and then try plaintext if something goes wrong with the push or pull. It doesn't make sense to try plaintext if a HTTP request went through while using TLS. This commit changes the logic to keep track of host/port combinations where a TLS attempt managed to do at least one HTTP request (whether the response code indicated success or not). If the host/port responded to a HTTP using TLS, we won't try to make plaintext HTTP requests to it.

This will result in better error messages, which sometimes ended up showing the result of the plaintext attempt, like this:

```
Error response from daemon: Get
http://myregistrydomain.com:5000/v2/: malformed HTTP response
"\x15\x03\x01\x00\x02\x02"
```

This fixes one of the things reported in https://github.com/docker/docker/issues/19834.

cc @docker/distribution-maintainers

Commit:
Smarter push/pull TLS fallback

With the --insecure-registry daemon option (or talking to a registry on
a local IP), the daemon will first try TLS, and then try plaintext if
something goes wrong with the push or pull. It doesn't make sense to try
plaintext if a HTTP request went through while using TLS. This commit
changes the logic to keep track of host/port combinations where a TLS
attempt managed to do at least one HTTP request (whether the response
code indicated success or not). If the host/port responded to a HTTP
using TLS, we won't try to make plaintext HTTP requests to it.

This will result in better error messages, which sometimes ended up
showing the result of the plaintext attempt, like this:

    Error response from daemon: Get
    http://myregistrydomain.com:5000/v2/: malformed HTTP response
    "\x15\x03\x01\x00\x02\x02"

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14926/console):

~~~console

---
1.0-dev/validate-lint)
Errors from golint:
distribution/pull.go:117:26: should omit type map[string]struct{} from declaration of var confirmedTLSRegistries; it will be inferred from the right-hand side
distribution/push.go:127:26: should omit type map[string]struct{} from declaration of var confirmedTLSRegistries; it will be inferred from the right-hand side

Please fix the above errors. You can test via "golint" and commit the result.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson780573791668478731.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23715/console):

~~~console

---
1.0-dev/validate-lint)
Errors from golint:
distribution/pull.go:117:26: should omit type map[string]struct{} from declaration of var confirmedTLSRegistries; it will be inferred from the right-hand side
distribution/push.go:127:26: should omit type map[string]struct{} from declaration of var confirmedTLSRegistries; it will be inferred from the right-hand side

Please fix the above errors. You can test via "golint" and commit the result.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson3703849982510422298.sh
+ set +x


---

~~~

vikstrous commented:
Sweet. I just tried this. It seems to work as advertised. It also fixes https://github.com/docker/docker/issues/20104 I think

cpuguy83 commented:
Just setting to design review since it's not ready and we don't have a better label right now.

vikstrous commented:
One more awkward thing is left. It's still not obvious what's going on if you try to push to a server that isn't listening on any ports. It looks like the connection on port 80 faild, but really the issue is that port 443 failed. Can we preserve the error message from the first failure?

```
$ docker push loc2.loc/admin/repo
The push refers to a repository [loc2.loc/admin/repo]
Put http://loc2.loc/v1/repositories/admin/repo/: dial tcp 192.168.1.2:80: getsockopt: connection refused
```

tiborvass commented:
Agree with https://github.com/docker/docker/pull/20253#issuecomment-183153277 otherwise this is a really welcome improvement.

runcom commented:
apart from https://github.com/docker/docker/pull/20253#issuecomment-183153277 this is fixing my issue :)

aaronlehmann commented:
Rebased; removed WIP note.

The bit about servers that aren't listening on either port is interesting. For much of the 1.10 dev cycle, we displayed the errors from all the attempts. This was found to be confusing in some cases and the change was reverted before the 1.10 release. See https://github.com/docker/docker/pull/19424

I'd like to find a solution to this, but I think it's better for this to be a separate PR, since it's more about error reporting than controlling the actual connection attempts. Could you please file an issue for tracking purposes?

stevvooe commented on distribution/push_v2.go:
```diff
@@ -61,15 +61,24 @@ type pushState struct {
+			return fallbackError{
+				err:         err,
+				confirmedV2: p.pushState.confirmedV2,
+				transportOK: true,
```
Should this use the variable value from above?

aaronlehmann replied:
> Should this use the variable value from above?

In the original version, `NewV2Repository` always returned `transportOK == true` on success, because completing this function successfully involves a successful HTTP "ping". Per your other comment, I've updated it to return `transportOK` as part of the error it returns. This should make it clearer that `transportOK` is only false in certain cases where `NewV2Repository` fails.

stevvooe replied:
Much cleaner.

stevvooe commented on distribution/registry.go:
```diff
@@ -60,14 +60,17 @@ func NewV2Repository(ctx context.Context, repoInfo *registry.RepositoryInfo, end
 	req, err := http.NewRequest("GET", endpointStr, nil)
 	if err != nil {
-		return nil, false, err
+		return nil, false, false, err
```
Is there anyway we can embed this in the error type? I am reticent in expanding the number of return values from this function.

Let me know if you need ideas.

aaronlehmann replied:
@stevvooe: I've updated the PR to use a different approach, please tell me if you think it's an improvement. Now `NewV2Repository` returns a `fallbackError` directly instead of depending on the caller to wrap the error it returns. This means we can skip the extra return value.

stevvooe commented:
LGTM

dmcgowan commented:
LGTM

thaJeztah commented:
moving to code review (@stevvooe @dmcgowan pls re-LGTM of code is ok; not sure if your LGTM's were on design or code)

calavera commented on distribution/push.go:
```diff
@@ -127,6 +133,19 @@ func Push(ctx context.Context, ref reference.Named, imagePushConfig *ImagePushCo
 			continue
 		}
 
+		parsedURL, urlErr := url.Parse(endpoint.URL)
```
can we extract this logic to a function? it looks like it's duplicated in `pull.go`.

aaronlehmann replied:
@calavera: This is a fair point - quite possibly one I would raise myself if I reviewed similar code. I thought about this while working on the PR, but didn't immediately see a clean way to split the code out.

The new code here involves parsing the URL and also maintaining a map. The direct way to reuse it would be to provide a function that parses the URL, checks for it in the map, and returns the parsed representation, and also a function that adds the URL to the map if appropriate. Both functions would have to take a fair amount of state in arguments or a special structure, and it seemed like it would be ugly.

The `Push` and `Pull` functions are nearly identical, except one calls `newPuller`/`Pull` and the other calls `newPusher`/`Push` (and they have log messages tailored to the particular operation). Ideally `Push` and `Pull` would be combined into something more generic. The obstacle is that `Pusher` and `Puller` are different interfaces, so I don't see an obvious way to write a clean generic version.

I will experiment with this and get back to you.

stevvooe replied:
@calavera Do you mean something like `RequiresTLS(url) (bool, error)`?

aaronlehmann replied:
@calavera: Do you prefer this approach? https://gist.github.com/aaronlehmann/dcec73055470ba81cd4f

I'm not sure I like the set of function arguments and return values:

    func checkRegistryIncompatibility(endpoint registry.APIEndpoint, confirmedTLSRegistries map[string]struct{}, confirmedV2 bool) (*url.URL, error) {

...but I don't really have strong feelings about this.

stevvooe replied:
I don't really think that adds clarity. The current code parses a url and checks the scheme. This is very clear. No need to be DRY for DRY's sake.

calavera replied:
@aaronlehmann 

https://play.golang.org/p/IBM0WZkHV1

aaronlehmann replied:
@calavera: I guess there would also be a method to mark a registry as a confirmed TLS registry? Where should the URL parsing happen for this? Should both methods have their own redundant calls to `url.Parse`?

calavera replied:
> Where should the URL parsing happen for this?

I really don't know why `APIEndpoint` stores `URL` as `string` and `Endpoint` stores `URL` as `*url.URL` but that looks like a smell tbh. Why don't we have a single place where all the parsing happens? Would that make all this logic easier?

aaronlehmann replied:
@calavera: I agree that it would make more sense for `APIEndpoint` to contain a parsed URL. I added a commit that makes this change.

Please let me know whether:

a) the change looks good
b) this should be broken out into a separate PR

Commit:
Change APIEndpoint to contain the URL in a parsed format

This allows easier URL handling in code that uses APIEndpoint.
If we continued to store the URL unparsed, it would require redundant
parsing whenver we want to extract information from it. Also, parsing
the URL earlier should give improve validation.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21871/console):

~~~console

---
2 ++++++++++++++++----
 16 files changed, 173 insertions(+), 50 deletions(-)
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson6376408117767320349.sh
+ set +x
INFO: Started at Thu Feb 18 00:55:50 CUT 2016. Script version 4-Jan-2016 15:19 PST
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson807757082628700816.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1147/console):

~~~console

---
8T00:57:34Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End of cleanup
INFO: Ended at Thu Feb 18 00:57:34 CUT 2016 (1m 46s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\jenkins\AppData\Local\Temp\hudson9092981600035931237.sh
+ set +e
+ set +x
INFO: Nuking /d/CI
time="2016-02-18T00:57:43Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI 
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23883/console):

~~~console

---
/cross/freebsd/arm/docker-1.11.0-dev
Created binary: bundles/1.11.0-dev/cross/freebsd/arm/docker-1.11.0-dev
Building: bundles/1.11.0-dev/cross/windows/amd64/docker-1.11.0-dev.exe
# github.com/docker/docker/registry
registry/config_windows.go:15: const initializer &url.URL literal is not a constant
registry/config_windows.go:22: undefined: net in net.URL
registry/config_windows.go:25: const initializer composite literal is not a constant
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson39307702116217026.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21874/console):

~~~console

---
2 ++++++++++++++++----
 16 files changed, 174 insertions(+), 51 deletions(-)
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson7589133441267409422.sh
+ set +x
INFO: Started at Thu Feb 18 01:48:59 CUT 2016. Script version 4-Jan-2016 15:19 PST
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson8661200936975952449.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21972/console):

~~~console

---


Ping statistics for 10.0.0.13:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 1ms, Average = 0ms
ERROR: Failed with exitcode 1 at Fri Feb 19 17:17:22 CUT 2016.
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson1343291185999005467.sh
+ set +x


---

~~~

tiborvass commented:
LGTM, the win2lin issue is known and being worked on.
